### ISSUE TITLE
Exception when using BinaryFormatter to deserialise entities with initialized proxies in associations

### ISSUE THREAD

**Issue by dimako** [nhibernate/nhibernate-core#2673]
I've created a little test project which runs on .NET Framework 4.8 and using NHibernate 5.3.5 which throws:
```
System.ArgumentNullException: 'Value cannot be null.
   at System.Collections.Generic.HashSet`1..ctor(IEnumerable`1 collection, IEqualityComparer`1 comparer)
   at NHibernate.Proxy.NHibernateProxyFactoryInfo.CreateProxyFactory()
   at NHibernate.Proxy.NHibernateProxyObjectReference.GetRealObject(StreamingContext context)
   at System.Runtime.Serialization.ObjectManager.ResolveObjectReference(ObjectHolder holder)
   at System.Runtime.Serialization.ObjectManager.DoFixups()
   at System.Runtime.Serialization.Formatters.Binary.ObjectReader.Deserialize(HeaderHandler handler, __BinaryParser serParser, Boolean fCheck, Boolean isCrossAppDomain, IMethodCallMessage methodCallMessage)
   at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Deserialize(Stream serializationStream, HeaderHandler handler, Boolean fCheck, Boolean isCrossAppDomain, IMethodCallMessage methodCallMessage)
   at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Deserialize(Stream serializationStream)
```
The database design is as follows:
```
create table [Role]
(
    RoleID int,
    [Name] nvarchar(50),
    PRIMARY KEY (RoleID)
)

create table [Resource]
(
    ResourceID int,
    [Name] nvarchar(50),
    RoleID int,
    ManagerID int,
    PRIMARY KEY (ResourceID),
    FOREIGN KEY (RoleID) REFERENCES [Role](RoleID),
    FOREIGN KEY (ManagerID) REFERENCES [Resource](ResourceID)
)

insert [Role] (RoleID, [Name])
values (1, 'role1')
insert [Role] (RoleID, [Name])
values (2, 'role2')

insert [Resource] (ResourceID, [Name], RoleID, ManagerID)
values (1, 'res1', 1, null)

insert [Resource] (ResourceID, [Name], RoleID, ManagerID)
values (2, 'res2', 2, 1)
```
There are two classes:
using System;

```
namespace NHibernate.Bug
{
    [Serializable]
    public class Role
    {
        public virtual string Name { get; set; }
        public virtual int Key { get; set; }
    }
}

using System;

namespace NHibernate.Bug
{
    [Serializable]
    public class Resource
    {
        public virtual int Key { get; set; }
        public virtual Resource Manager { get; set; }        
        public virtual string Name { get; set; }
        public virtual Role ResourceRole { get; set; }
    }
}
```
and their mappings are:
```
<?xml version="1.0"?>
<hibernate-mapping xmlns="urn:nhibernate-mapping-2.2" assembly="NHibernate.Bug"
    namespace="NHibernate.Bug">

  <class name="Role" table="Role">
    <id name="Key">
      <column name="RoleID" sql-type="int" not-null="true"/>
      <generator class="assigned" />
    </id>
    <property name="Name"/>
  </class>

</hibernate-mapping>

<?xml version="1.0"?>
<hibernate-mapping xmlns="urn:nhibernate-mapping-2.2" assembly="NHibernate.Bug"
    namespace="NHibernate.Bug">

  <class name="Resource" table="Resource" dynamic-insert="true">
    <id name="Key" column="ResourceID" >
      <generator class="assigned" />
    </id>
    <property name="Name"/>
    <many-to-one name="Manager" class="Resource" column="ManagerID"/>
   
    <many-to-one name="ResourceRole" class="Role" column="RoleID"/>
  </class>

</hibernate-mapping>
```

The code that is failing is:
```
var session = CreateSessionFactory().OpenSession();

            // if we don't run this line the test passes
            var resource1 = session.CreateCriteria<Resource>("r").Add(Restrictions.Eq("r.Key", 2)).UniqueResult<Resource>();

            var resource2 = session.QueryOver<Resource>()
                .Where(res => res.Key == 2)
                .Fetch(SelectMode.Fetch, res => res.Manager)
                .SingleOrDefault();

            var serialised = Serialize(resource2);
            var deserialised = Deserialize(serialised);
```

For some reason it only fails if we execute the criteria query before the queryover one.
Please see below the full fixture code:

```
using Microsoft.VisualStudio.TestTools.UnitTesting;
using NHibernate.Cfg;
using NHibernate.Criterion;
using System.IO;
using System.Runtime.Serialization.Formatters.Binary;

namespace NHibernate.Bug
{
    [TestClass]
    public class ProxyFixture
    {
        [TestMethod]
        public void ItShouldSerializeAndDeserialize()
        {
            var session = CreateSessionFactory().OpenSession();

            // if we don't run this line the test passes
            var resource1 = session.CreateCriteria<Resource>("r").Add(Restrictions.Eq("r.Key", 2)).UniqueResult<Resource>();

            var resource2 = session.QueryOver<Resource>()
                .Where(res => res.Key == 2)
                .Fetch(SelectMode.Fetch, res => res.Manager)
                .SingleOrDefault();

            var serialised = Serialize(resource2);
            var deserialised = Deserialize(serialised);
        }

        private ISessionFactory CreateSessionFactory()
        {
            var config = new Configuration();
            config.SetProperty(Environment.ConnectionString, "Server=xxx;Database=xxx;User Id=xxx;Password=xxx;");
            config.SetProperty(Environment.ConnectionDriver, "NHibernate.Driver.SqlClientDriver");
            config.SetProperty(Environment.Dialect, "NHibernate.Dialect.MsSql2012Dialect, NHibernate");
            config.SetProperty(Environment.ShowSql, "false");
            config.SetProperty(Environment.DefaultFlushMode, "Commit");
            config.SetProperty(Environment.FormatSql, "false");
            config.SetProperty(Environment.BatchSize, "50");
            config.AddAssembly(GetType().Assembly);              

            return config.BuildSessionFactory();
        }

        protected byte[] Serialize(object value)
        {                     
            var formatter = new BinaryFormatter();
            using (var stream = new MemoryStream())
            {
                formatter.Serialize(stream, value);
                var result = new byte[stream.Length];
                stream.Position = 0;
                stream.Read(result, 0, (int)stream.Length);
                stream.Close();
                return result;
            }
        }

        protected object Deserialize(byte[] state)
        {
            using (MemoryStream stream = new MemoryStream())
            {
                stream.Write(state, 0, state.Length);
                stream.Position = 0;
                var formatter = new BinaryFormatter();
                return formatter.Deserialize(stream);
            }
        }
    }
}

```

**Comment by bahusoid** [comment]
> that have got lazy loaded properties

I see no lazy properties in your mappings so I changed the title to avoid confusion.

**Comment by dimako** [comment]
> > that have got lazy loaded properties
> 
> I see no lazy properties in your mappings so I changed the title to avoid confusion.

I apologise for the confusion but we don't specifically mark properties with lazy="false" in the mapping are they not lazy loaded?

**Comment by gliljas** [comment]
Properties are not lazy by default. Reference properties are.

**Comment by dimako** [comment]
> Properties are not lazy by default. Reference properties are.

Of course! I assumed it was to do with lazy loading because I think the problem is in Role property. It is not referenced directly but I think it’s being lazy loaded during deserialization and that’s why everything is falling over. But most likely my assumption is wrong and I don’t want to mislead anyone. As you can see from the data the Resource and the Manager have different Roles. If we update them to have the same Role the issues goes away. That’s another reason I thought it had something to do with lazy loading .

**Comment by fredericDelaporte** [comment]
Isn't that #2563?

**Comment by dimako** [comment]
> Isn't that #2563?

I did initially added a comment there but somebody pointed out that it looked a different issue. And the exception looks different

**Comment by dimako** [comment]
It seems the issue was caused by this commit https://github.com/nhibernate/nhibernate-core/pull/1968/commits/72b7eb4ab22506796287e715abd96268bb3839f4
Hope that is helpful. Really looking forward to that being resolved!

Thank you!

**Comment by bahusoid** [comment]
Does it mean it works in 5.2?

**Comment by dimako** [comment]
> Does it mean it works in 5.2?

Yes looks like it

**Comment by fredericDelaporte** [comment]
> Yes looks like it does

You mean, you think it works correctly, but have not ascertain it? For me it was not supposed to work, but maybe in a different way, like no exception on deserialization but badly initialized lazy properties.

**Comment by dimako** [comment]
> > Yes looks like it does
> 
> You mean, you think it works correctly, but have not ascertain it? For me it was not supposed to work, but maybe in a different way, like no exception on deserialization but badly initialized lazy properties.

It looks it was working in 5.2 but got broken after the mentioned commit was pushed in.

**Comment by bahusoid** [comment]
> but badly initialized lazy properties

There is no lazy properties

**Comment by maca88** [comment]
> It seems the issue was caused by #1968

That is correct, but unfortunately I was not able to fix it without removing part of the optimization made in #1968. In case anyone knows a better fix, feel free to suggest/apply the alternative to #2719.

Off topic:
When regenerating the async code I spotted that one comment from #2713 wasn't updated in the async file.

**Comment by bahusoid** [comment]
> In case anyone knows a better fix

See attempt to save your optimizations in #2720. I also wasn't able to reproduce this issue for `FieldInterceptorObjectReference` so didn't change it...

**Comment by fredericDelaporte** [comment]
Fixed by #2720.
@dimako, may you check that the second 5.3.8 [MyGet build](https://www.myget.org/feed/nhibernate/package/nuget/NHibernate), once available, does work for you? (It should be available in a couple of hours at most.)

**Comment by dimako** [comment]
> Fixed by #2720.
> @dimako, may you check that the second 5.3.8 [MyGet build](https://www.myget.org/feed/nhibernate/package/nuget/NHibernate/5.3.8-dev.3328) does work for you?

Thanks a lot, the bug seems to be fixed now!
